gobblin-utility/src/main/java/org/apache/gobblin/util/ConfigUtils.java

```diff
@@ -116,7 +116,8 @@ public static Properties configToProperties(Config config, Optional<String> pref
       Config resolvedConfig = config.resolve();
       for (Map.Entry<String, ConfigValue> entry : resolvedConfig.entrySet()) {
         if (!prefix.isPresent() || entry.getKey().startsWith(prefix.get())) {
-          String propKey = desanitizeKey(entry.getKey());
+          // Intern the string so that constant keys are not duplicated to save memory
+          String propKey = desanitizeKey(entry.getKey()).intern();
```

Comment:
So is this the place where we populate configToProperties in FlowSpec? Should we `intern()` the values as well.Fliptop datapacks has duplicate values as well. I can give examples on slack because these are internal configs

Reply:
I would be scared to intern() the values, if unless we can guarantee that the values are acting immutable here. Otherwise they will point to the same object, and any config value that is changed will change across all uses.

Reply:
propVal is a String and Strings are immutable, right? When a config value is changed, it does not change the value, it creates a new value in the underlying map and set the new value to that key, right?

Reply:
Ah you're right it would create a new one, I misunderstood some of the underlying risks of intern() 😄 . Doing some more reading, it does have some risks if there are a number of unique values though. Given that there are many values that are unique in key vals for configs, it might lead to other issues moreso than keys, like affecting the permanent memory pool size.

Reply:
The exact behavior of intern() is not well document, and changed across java versions. Here is an article where a guy reverse-engineers it - http://java-performance.info/string-intern-in-java-6-7-8

Reply:
The article Alex cited says to "Use -XX:StringTableSize JVM parameter in Java 7 and 8 to set the string pool map size." Is that not needed here?

Reply:
@umustafi thanks for the callout, it defaults to ~60k strings which is likely more than enough for keys (as we have a fixed number of configuration keys that should not exceed that number) but I will add the argument `-XX:+PrintStringTableStatistics` so that I can see how many we're actually using. These arguments would be in the service deployable runtime args, so it wouldn't be included in this repository.

Reply:
@Will-Lo This method is invoked in a number of places and not limited to Gobblin service. One concern is that the number of keys may not be as small as you imagine. In production settings, Gobblin jobs could load system configs set by AZ/Hadoop or even the host where a job/container is running. These can be non-trivial in number. 60k may be sufficient. But what happens if the number of config keys exceed this value? Will the JVM just crash?

Reply:
It may be safer to introduce an overloaded method with an additional boolean flag as an argument, which when enabled interns the keys/values. FlowSpec builder can then call this overloaded method and the rest of the usage can remain as-is.